docs/rfc/rfc-002-ipc-ecosystem.md

```diff
@@ -0,0 +1,350 @@
+but that is a larger project since it adds P2P communication load, and might
+thus have an impact on the performance of consensus.
+
+If event subscription can be moved into the P2P network, we could entirely
+remove the websocket transport, even for clients that still need access to the
+RPC service. Until then, we may still be able to reduce the scope of the
+websocket endpoint to _only_ event subscription, by moving uses of the RPC
+server as a proxy to ABCI over to the gRPC interface.
+
+Having the RPC server still makes sense for local bootstrapping and operations,
+but can be further simplified. Here are some specific proposals:
+
+- Remove the HTTP GET interface entirely.
+- Simplify JSON-RPC plumbing to remove unnecessary reflection and wrapping.
+- Remove the gRPC interface (this is already planned for v0.36).
+- Separate the websocket interface from the rest of the RPC service, and
+  restrict it to only event subscription.
+  - Eventually: Remove the websocket interface entirely.
+
+These changes would preserve the ability of operators to issue queries with
+curl (but would require using JSON-RPC instead of URI parameters). That would
+be a little less user-friendly, but for a use case that should not be that
+prevalent.
+
+These changes would also preserve compatibility with existing JSON-RPC based
+code paths like the `tendermint` CLI and the light client (even ahead of
+further work to remove that dependency).
+
+**Design goal:** An operator should be able to disable non-local access to the
+RPC server on any node in the network without impairing the ability of the
+network to function for service of state replication, including light clients.
+
+**Design principle:** All communication required to implement and monitor the
+consensus network should use P2P, including the various synchronizations.
+
+### Options for ABCI Transport
+
+The majority of current usage is in Go, and the majority of that is mediated by
+the Cosmos SDK, which uses the "direct call" interface. There is probably some
+opportunity to clean up the implementation of that code, notably by inverting
+which interface is at the "top" of the abstraction stack (currently it acts
+like an RPC interface, and escape-hatches into the direct call). However, this
+general approach works fine and doesn't need to be fundamentally changed.
+
+For applications _not_ written in Go, the two remaining options are the
+"socket" protocol (another variation on varint-prefiexed protobuf messages over
+an unstructured stream) and gRPC. It would be nice if we could get rid of one
+of these to reduce (unneeded?) optionality, and the "socket" protocol is the
+most obvious choice. gRPC is more complex, but is already widely used in the
+rest of the ecosystem (including the Cosmos SDK).
+
+The main question is whether requiring gRPC imposes an undue burden on the
+authors of applications in languages without good gRPC support. I propose that
```

Comment:
Probably worth noting that our socket protocol implementation does require that a language support serializing protobufs for use on the wire. The set of languages that support protocol buffers but _not_ gRPC is likely not very large.

The [protocol buffer](https://github.com/protocolbuffers/protobuf#protobuf-runtime-installation) repo lists the follow languages as officially supported (although there may be others):

C++ (include C++ runtime and protoc) 
Java 
Python
Objective-C 
C# 
JavaScript 
Ruby 
Go
PHP 
Dart

[gRPC's documentation lists](https://grpc.io/docs/languages/):

C#
C++
Dart
Go
Java
Kotlin
Node
Objective-C
PHP
Python
Ruby

These appear to be roughly the same list with the addition of kotlin, which likely is achieved 'for free' as a JVM language.

Reply:
Notably, there is a third party list [in the protobuf repo](https://github.com/protocolbuffers/protobuf/blob/master/docs/third_party.md) that shows some third party support for protobuf in languages where gRPC is not supported (i.e. scala, lua, julia).

Reply:
Yes, agreed. That is more or less the point I was trying to make here in this paragraph. I mentioned above that the socket protocol also uses protobuf, but I welcome suggestions for how to make it more clear.

Reply:
Thanks, that helps. I didn't quite adopt that wording exactly, but it helped me clarify the text.